apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -190,10 +191,66 @@ def __init__(self):
             ofcst.AGENT_TYPE_OPFLEX_OVS)
         ha_ip_db.HAIPOwnerDbMixin.__init__(self)
 
+    def bind_port(self, context):
+        """Get port binding per host.
+
+           Overriding the superclass implementation
+           in order to support multiple agent types
+           in a single mechanism driver (DVS and OpFlex)
+        """
+
+        LOG.debug("Attempting to bind port %(port)s on "
+                  "network %(network)s",
+                  {'port': context.current['id'],
+                   'network': context.network.current['id']})
+        vnic_type = context.current.get(portbindings.VNIC_TYPE,
+                                        portbindings.VNIC_NORMAL)
+        if vnic_type not in self.supported_vnic_types:
+            LOG.debug("Refusing to bind due to unsupported vnic_type: %s",
+                      vnic_type)
+            return
+
+        agent_list = context.host_agents(self.agent_type)
```

Comment:
As I discussed with Thomas, I think we want to look at the port's device_owner and see if its nova. If it is, then try binding with AGENT_TYPE_DVS. If not, or if that fails (i.e. no DVS agent on the node), then try binding with AGENT_TYPE_OPFLEX_DVS. I would do these as separate loops rather than by extending the list of agents as you have.